src/atom_types.F

```diff
@@ -1157,6 +1180,166 @@ SUBROUTINE release_atom_orbs(orbs)
+                                                                *abscissa(i)*atom%basis%grid%rad(j), nu)*SQRT(weights(i))
+                     END DO
+                  END DO
+               END DO
+            ELSE
+               ! Setup kernel for Ewald operator
+               ! Because of the high computational costs of its calculation, we precalculate it here
+               ! Choose it symmetric to further reduce the costs
+               nr = atom%basis%grid%nr
+               ALLOCATE (atom%hfx_pot%kernel(nr, nr, 0:atom%state%maxl_calc + atom%state%maxl_occ))
+               atom%hfx_pot%kernel = 0.0_dp
+               DO nu = 0, atom%state%maxl_calc + atom%state%maxl_occ
+                  DO i = 1, nr
+                     DO j = 1, i
+                        atom%hfx_pot%kernel(j, i, nu) = bessel0(2.0_dp*atom%hfx_pot%omega &
+                                                                *atom%basis%grid%rad(i)*atom%basis%grid%rad(j), nu)
+                     END DO
+                  END DO
+               END DO
+            END IF
+         END IF
+      ELSE
+         NULLIFY (xc_section)
+         do_hfx = .FALSE.
+      END IF
+
+   END SUBROUTINE setup_hf_section
+
+! **************************************************************************************************
+!> \brief ...
+!> \param abscissa ...
+!> \param weights ...
+!> \param nn ...
+! **************************************************************************************************
+   SUBROUTINE get_gauss_hermite_weights(abscissa, weights, nn)
+      REAL(KIND=dp), DIMENSION(:), INTENT(OUT)           :: abscissa, weights
+      INTEGER, INTENT(IN)                                :: nn
+
+      INTEGER                                            :: counter, ii, info, liwork, lwork
+      INTEGER, ALLOCATABLE, DIMENSION(:)                 :: iwork
+      REAL(KIND=dp), ALLOCATABLE, DIMENSION(:)           :: diag, subdiag, work
+      REAL(KIND=dp), ALLOCATABLE, DIMENSION(:, :)        :: eigenvec
+
+      ALLOCATE (work(1), iwork(1), diag(2*nn), subdiag(2*nn - 1), eigenvec(2*nn, 2*nn))
+      lwork = -1
+      liwork = -1
+      diag = 0.0_dp
+      DO ii = 1, 2*nn - 1
+         subdiag(ii) = SQRT(REAL(ii, KIND=dp)/2.0_dp)
+      END DO
+
+      CALL DSTEVD('V', 2*nn, diag, subdiag, eigenvec, 2*nn, work, lwork, iwork, liwork, info)
+      CPASSERT(info == 0)
+
+      lwork = INT(work(1))
+      liwork = iwork(1)
+      DEALLOCATE (work, iwork)
+      ALLOCATE (work(lwork), iwork(liwork))
+
+      CALL DSTEVD('V', 2*nn, diag, subdiag, eigenvec, 2*nn, work, lwork, iwork, liwork, info)
+      CPASSERT(info == 0)
+
+      DEALLOCATE (work, iwork, subdiag)
+
+      counter = 0
+      DO ii = 1, 2*nn
+         IF (diag(ii) > 0.0_dp) THEN
+            counter = counter + 1
+            abscissa(counter) = diag(ii)
+            weights(counter) = rootpi*eigenvec(1, ii)**2
+         END IF
+      END DO
+      CPASSERT(counter == nn)
```

Comment:
I'd probably have used something like `CPASSERT(ALL(diag > 0.0_dp))` prior to the loop to separate error-handling and algorithm, but I guess that's just style. But what I really miss for these kind of checks is a small comment giving the developer/user some ideas on what went wrong on how to fix it.

Reply:
I will add more comments about that! CPASSERT(ALL(diag > 0.0_dp)) is not possible because the eigenvalues will be symmetric wrt the origin in this case and I will need only the positive roots. (This algorithm calculates the roots of the Hermite polynomials that are even or odd functions.)

Reply:
ah, I see, sorry. But yes, please a comment.